[13.0] Deprecate job decorators

guewen opened the pull request:
`@job` and `@related_action` are still supported but emit warnings and
will be removed at some point (code to remove has a TODO with:job-no-decorator:)

The initial reasons for this change are in #273, but it also brings nice side effects:

* the options previously set by the decorators are now stored on
  "queue.job.function" records, which can be explicitly created
  as XML records, which mean they can now be **edited by users**
* **any method can be used with "with_delay()"**, if no "queue.job.function"
  exists, default values are used (root channel, default related
  action, ...)

This one will require a major bump.

My wish is:

- 12.0: deprecate decorators but show no warning when they are used
- 13.0: deprecate decorators and show warnings when they are used
- 14.0: entirely remove them, as we are starting migrations, it will be easy to tackle in the modules

Remaining tasks:

* [x] Adapt tests
* [x] Document the configuration (a lot of documentation is in the decorators, move it elsewhere + in readme)
* [x] Figure out how to handle duplicates `queue.job.function` (not big issue though)
* [x] update other modules of the repository
* [x] As any method can be delayed, maybe we should guard create/write (on a subset of fields) on `queue.job` from RPC (OTOH only queue job managers can do it, but at the very least they should not be able to create a raw queue.job record for a method starting with `_`
* [x] More checks/constraints on config, improve views, ...
* [x] Add model / method fields on `queue.job.function` and compute the name from there

Fixes #273

Commit:
Deprecate job decorators

* @job and @related_action are still supported but emit warnings and
  will be removed at some point (code to remove has a TODO with
  :job-no-decorator:)
* the options previously set by the decorators are now stored on
  "queue.job.function" records, which can be explicitly created
  as XML records (and then editable!)
* any method can be used with "with_delay()", if no "queue.job.function"
  exists, default values are used (root channel, default related
  action, ...)

Commit:
Rebind job functions/channels with same name during install

As previously, the job functions and channels were created automatically
by "_register_job" on methods decorated by `@job`, databases have
records without xmlids.

To prevent duplicates (or unique constraint error) in case of channels,
the create method automatically merges the new record with the existing
record. This behavior happens only when installing/upgrading a module,
records created manually behave normally.

If 2 modules adds the same channel or job function, they'll be merged
together with 2 xmlids, on uninstallation of one of the module, the
record will be kept thanks to the second xmlid.

Commit:
Update documentation for deprecation of decorators

Commit:
queue_job_cron: remove `@job` decorator

Commit:
base_import_async: remove `@job` decorator

Commit:
Add safety to prevent crafting queue.job records

As we can delay a job on any method, and the queue.job model is
accessible from RPC (as any model), prevent to:

* create a queue.job using RPC
* write on protected fields (e.g. method name) using RPC

Admittedly, the risk is low since users need have Queue Job Manager
access to create/write on jobs, but it would allow these users to call
internal methods.

The check is done using a context key that must be equal to a sentinel
object, which is impossible to pass through RPC.

Commit:
Add missing pylint check "development-status-allowed"

Commit:
Fix invalid development_status (accepted value is Production/Stable)

guewen commented:
No runbot build?

lmignon commented on base_import_async/data/queue_job_function_data.xml:
```diff
@@ -0,0 +1,19 @@
+        model="queue.job.function"
+    >
+        <field name="name"><![CDATA[<base_import.import>._import_one_chunk]]></field>
+        <field
```
@guewen Wouldn't it be easier to have two fields: model and method_name and to compute the name from these two fields? This would allow a more natural syntax.

guewen replied:
Yes for sure! I add this to the remaining tasks

guewen replied:
Added in https://github.com/OCA/queue/pull/274/commits/27687b2ed9009edfd7f4baeb4dce0ebcdf025901

Commit:
Replace job function name by model + method fields

A many2one to select the model + a char for the method name is more
convenient and error-proof than having to type the expected format
correctly.

sebalix reviewed (approved):
LG,
I suppose the website https://odoo-connector.com/  will need an update later :) (but it seems related to https://github.com/OCA/connector/tree/13.0/connector )

guewen commented:
> I suppose the website https://odoo-connector.com/ will need an update later :)

It is supposed to be refreshed by travis, but IIRC triggered by new commits in OCA/connector

guewen commented:
> I suppose the website https://odoo-connector.com/ will need an update later :)

But of course it doesn't include the readme's content. We need a global refresh of the doc, ideally its own one.

sebastienbeau commented:
:+1: Really good change ! LGTM code review

Commit:
Add missing sudo() when reading queue.job.function

Users shouldn't need access to queue.job.function.
Read config as sudo().

guewen commented:
@lmignon could you update your review please? :)

lmignon reviewed (approved):
Thank you @guewen

guewen commented:
/ocabot merge major

rousseldenis commented on queue_job/job.py:
```diff
@@ -810,6 +832,25 @@ def retryable_example():
+        '<record id="job_function_[insert model]_{method}"'
+        ' model="queue.job.function">\n' + "\n".join(xml_fields) + "\n</record>"
+    ).format(**{"method": func.__name__, "retry_pattern": retry_pattern})
+    _logger.warning(
```
@guewen IMHO this is too much as it will break all runbot builds in repos that use queues. See https://runbot.odoo-community.org/runbot/build/3450323 for example.

Is it wanted ? Or maybe this can go into INFO.

guewen replied:
You are right, I didn't realize runbot would not agree. We should move it to info, and move the XML snippet to debug because it is too verbose at the moment. I'll try to open a PR tomorrow.
